keps/sig-windows/1981-windows-privileged-container-support/README.md

```diff
@@ -459,87 +464,136 @@ Windows privileged containers will be implemented with [Job Objects](https://doc
-	WindowsSandboxSecurityContext security_context = 1;
+Add WindowsPodSandboxConfig (and it to PodSandboxConfig)
+
+```protobuf
+message WindowsPodSandboxConfig {
+  WindowsSandboxSecurityContext security_context = 1;
 }
 ```
 
-A new WindowsSandboxSecurityContext:
-```
+Add WindowsSandboxSecurityContext:
+
+```protobuf
 message WindowsSandboxSecurityContext {
-	string run_as_user = 1;
-	bool privileged = 2;
+  string run_as_username = 1;
+  bool privileged = 2;
 }
 ```
 
-The new field for the WindowsContainerSecurityContext:
-```
+Update WindowsContainerSecurityContext by adding privileged field:
+
+```protobuf
 message WindowsContainerSecurityContext {
-	string run_as_username = 1;
-	string credential_spec = 2;
+  string run_as_username = 1;
+  string credential_spec = 2;
   bool privileged = 3;
 }
 ```
-### Kubelet Implementation Details
 
-#### <b>Privileged Flag</b>
-There are no kubernetes API changes required.  Kubelet can use the privileged flag from SecurityContext api and pass that to the new privileged flag in the CRI layer.  
+*Note:* For alpha annotations on RunPodSandbox and CreateContainer CRI calls may be used until a version of containerd with Windows privileged container support is released.
 
-Add functionality to Kuberuntime_sandbox to split out the linux sandbox creation and add windows sandbox creation (WIP PR). 
+### Kubernetes API updates
 
-https://github.com/kubernetes/kubernetes/blob/a9f1d72e1de6450b890a0c0e451725468f54f515/pkg/kubelet/kuberuntime/kuberuntime_sandbox.go#L136 
+#### Privileged Flag
 
-#### <b>Host Network Mode</b>
+A new boolean field named `privileged` will be added to [WindowsSecurityContextOptions](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#windowssecuritycontextoptions-v1-core).
 
-If using Windows privileged containers the Host Network Mode will also be enabled, as the pod will automatically get the host IP.  The PodSpec has a hostNetwork boolean field:
+On Windows, all containers in a pod must be privileged. Because of this behavior and because `WindowsSecurityContextOptions` already exists on both `PodSecurityContext` and `Container.SecurityContext` Windows containers will use this new field instead of solely relying on the existing `privileged` field which only exists on `SecurityContext`. Because many policy tools currently look for the existing `SecurityContext.privileged` field we will add validation to requrire this field also be set on all containers in a privileged pod.
 
-https://github.com/kubernetes/kubernetes/blob/d20fd4088476ec39c5ae2151b8fffaf0f4834418/pkg/kubelet/container/helpers.go#L227-L229
+API validation will ensure that if `privileged` is set in `Container.SecurityContext.WindowsSecurityContextOptions` for any container then then `privileged` must also set in `PodSecurityContext.WindowsSecurityContextOptions`. Current behavior applies pod `WindowsSecurityContextOptions` to all container level `WindowsSecurityContextOptions` if not specified.
 
```

Comment:
IIUC, this means that for a pod, there will be 3 different `privileged` fields:
* `PodSpec.PodSecurityContext.WindowsOptions.Privileged`
* `Container.SecurityContext.WindowsOptions.Privileged`
* `Container.SecurityContext.Privileged`.
This seems confusing to me.

Could we just reuse `Container.SecurityContext.Privileged`, so that `WindowsOptions.Privileged` is not needed?
Could we use a different word other than `Privileged`, e.g. `WindowsHostAccess`? 

I feel like either could be fine, because the `Privileged` word never has a clear definition for linux containers https://github.com/kubernetes/kubernetes/issues/44503. However, having 3 different `privileged` fields seems quite confusing to me.

BTW, on linux, `privileged` effectively means "disable all security features + allow accessing all devices". And in containerd, we even added a "privileged without host devices" option. https://github.com/containerd/cri/issues/1213

Reply:
Initially I was proposing to require `PodSpec.PodSecurityContext.WindowsOptions.Privileged` be set.
Current behavior applies `PodSpec.PodSecurityContext.WindowsOptions` to `Container.SecurityContext.WindowsOptions` for all containers (unless it is explicitly set).
I agree that having to set so many different fields will be cumbersome.

@liggitt wanted to see `Container.SecurityContext.Privileged` be set so existing policy tools would work.

I do not think we can re-use `Container.SecurityContext.Privileged` by itself because we would need to perform validation to ensure all containers in a pod are privileged - for Windows only - and we do not have a definitive way of knowing if a pod is intended to be scheduled to a Windows machine during API validation.

If we add a new `Container.SecurityContext.WindowsPrivileged` field we could require that to be set on all containers in a pod.

@liggitt, @Random-Liu how does this sound?

Also, Windows has no way of supporting a "privileged without host devices" options.
Windows privileges don't have any options for allowing access to devices unfortunately. (here are the available privileges for reference https://docs.microsoft.com/en-us/windows/win32/secauthz/privilege-constants)

Reply:
> we do not have a definitive way of knowing if a pod is intended to be scheduled to a Windows machine during API validation

If we think `Container.SecurityContext.Privileged` is a better API to use, maybe we can try to solve this problem instead?

For example, perform the check on the kubelet side. Because kubelet sits on the windows node, and has windows specific log. This would also address @yujuhong's comment above. https://github.com/kubernetes/enhancements/pull/2288#discussion_r564935794

And we can argue that this is only current implementation limitation, instead of API enforcement.

Reply:
> If we think `Container.SecurityContext.Privileged` is a better API to use, maybe we can try to solve this problem instead?

I'm not sure that using `Container.SecurityContext.Privileged` is a better API to use.
My biggest concern here is that privileged containers on Windows will work very differently than on Linux and re-using the privileged field will give users false impressions about their capabilities. 

I don't have super-strong feelings here and would definitely like input everyone.

The other thing to consider is that `WindowsOptions.RunAsUsername` will be the primary way to limit host access from privileged containers so it would be nice to have the two fields near each other in the pod-specs.

Reply:
> My biggest concern here is that privileged containers on Windows will work very differently than on Linux and re-using the privileged field will give users false impressions about their capabilities.

If that is the case, a new name might be more clear here.

`privileged` is convenient, but on the other hand, it is also confusing. It is hard to explain or for users to know what they can exactly do. We even discussed whether to remove it (at least from CRI) https://github.com/kubernetes/kubernetes/issues/44503#issuecomment-294197338.

Because of this, actually this is a good opportunity for us to at least fix this for Windows, instead of introducing another confusing field.

Can we come up with a better name than `Privileged` to better describe what users can do? `WindowsRawJob`, `WindowsHostAccess`, etc. I'm bad at naming.

Reply:
Thank you for the context and a link to the discussion about dropping the CRI field.
I would be OK coming up with a more descriptive name.
We discussed this internally too and didn't find a great name but we'll brainstorm some more.

I hope that if we have a sufficiently different field we would not need to require users also set `Container.SecurityOptions.Privileged`
I'll reach out to sig-auth to get their opinion as well.

Reply:
Thanks a lot! The overall workflow LGTM. The Kubernetes API field structure and naming is the only thing that I have concern with. And that decision would probably affect the CRI field naming as well.

To summarize my concern: The definition of `Privileged` field itself is not clear, having 3 of them in the pod spec, and require them to be set in a specific combination seems confusing to users. It would be ideal if we can avoid adding new `privileged` fields, either by reusing the existing `privileged` field, or add fields with new names (preferably with a more descriptive name).

Reply:
Thanks again!

I spoke with some other folks in MSFT and right now we are considering calling this `WindowsOptions.HostJob` in the kubernets API and `WindowsContainerSecurityContext.HostJobIsolation` in the CRI layer (to align with process / hyper-v isolation). We will also provide a lot of documentation around these these "privileged" containers.

I'll update the PR soon to capture a the context from this discussion then reach out to sig-auth / sig-api folks for another look.

As previously mentioned I think it would be a better user experience to have these new fields live on the same object as `RunAsUsername` since these fields will be used together.

Reply:
> I spoke with some other folks in MSFT and right now we are considering calling this WindowsOptions.HostJob in the kubernets API and WindowsContainerSecurityContext.HostJobIsolation in the CRI layer

The API sounds more clear to me. :)

>  (to align with process / hyper-v isolation).

This does make me wonder that, the `Job` isolation becomes a similar concept with the `Hyper-V` isolation - yet another isolation mode. And based on your doc, the new isolation mode requires container images to be built differently, and has different system compatibility requirement.

And I'm a bit torn that:
* On one hand, in the windows world, it could be modeled as yet another isolation mode, and we probably want to use a similar mechanism with `Hyper-V`. IIRC, we planned to support `Hyper-V` as a runtime class.
* On the other hand, if we compare the same use case and similar mechanism in the linux world, it is closer to `privileged` (shared host kernel with more privileges), so seem more like a container/pod config option than a separate runtime class.

If we plan to align the pod namespace, and potentially support privileged and non-privileged containers side by side in the future, this does look more like a special "container mode", instead of a "pod level runtime class". At the pod level, the "sandbox" is still the same thing.

Reply:
I was thinking about this more and I think this is more similar to a special "container mode" as you mentioned and less like an isolation mode.

Reply:
I updated the KEP to reflect these discussion points. PTAL.